### PR TITLE
US25988: Terraform UPDATE operation for load balancer and its configuration

### PR DESCRIPTION
**Pull request by Pushpalatha555**
**VALIDATED THE UPDATE OPERATION FOR LB AND ITS CONFIGURATIONS**

MON-CREATE
MON-UPDATE
POOL-CREATE
POOL-UPDATE
PROF-CREATE
PROF-UPDATE
VS-CREATE
VS-UPDATE

### CODE REVIEW

File: examples/data-sources/hpegl_vmaas_lb/data-source.tf [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# (C) Copyright 2021 Hewlett Packard Enterprise Development LP
+
+data "hpegl_vmaas_lb" "tf_lb" {
+  name         = 1
```
Comment (manjunath-batakurki) [inline examples/data-sources/hpegl_vmaas_lb/data-source.tf:4]: is this right name ?

File: examples/data-sources/hpegl_vmaas_lb_pool/data-source.tf [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# (C) Copyright 2021 Hewlett Packard Enterprise Development LP
+
+data "hpegl_vmaas_lb_pool" "tf_lb_pool" {
+  name         = 3
```
Comment (manjunath-batakurki) [inline examples/data-sources/hpegl_vmaas_lb_pool/data-source.tf:4]: ditto ?

File: internal/cmp/client.go [hunk 3]
```diff
@@ -40,6 +40,11 @@ type Client struct {
 	EdgeCluster               DataSource
 	TransportZone             DataSource
 	DSLoadBalancer            DataSource
+	DSLoadBalancerPool        DataSource
+	// DSLBMonitor               DataSource
+	// DSLBProfile               DataSource
```
Comment (manjunath-batakurki) [inline internal/cmp/client.go:45]: why are these commented ?

File: examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf [hunk 4]
```diff
@@ -1,15 +1,16 @@
...
+  name  = "tf_LB"  
   description  = "creating load balancer for test"
-  enabled      =    true        
-  visibility   = "public"
+  enabled      =    true      
   resource_permission {
     all = true
   }
   config {
+    log_level = "Info"
```
Comment (gandharvas) [inline examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf:11]: just a doubt, are these case sensitive?

File: internal/cmp/constants.go [hunk 5]
```diff
@@ -24,5 +24,19 @@ const (
 	tier1GatewayType             = "NSX-T Tier-1 Gateway"
 	routerFirewallExternalPolicy = "GatewayPolicy"
 
+	// load balancer consts
+	ApplicationProfile       = "application-profile"
+	ServiceTypeLBHttpProfile = "LBHttpProfile"
+	tier1                    = "/infra/tier-1s/26cdb82e-0057-4461-ad4d-cddd61d77b1f"
+	networkServerID          = 1
+	size                     = "SMALL"
```
Comment (gandharvas) [inline internal/cmp/constants.go:32]: Same question here, are these case sensitive? If yes, we need to make that it gets reflected in the documentation.

File: internal/cmp/lb_monitors.go [hunk 6]
```diff
@@ -24,31 +24,27 @@ func newLoadBalancerMonitor(loadBalancerClient *client.LoadBalancerAPIService) *
...
 
 func (lb *loadBalancerMonitor) Read(ctx context.Context, d *utils.Data, meta interface{}) error {
-	var lbMonitorResp models.GetSpecificLBMonitorResp
+	setMeta(meta, lb.lbClient.Client)
+	var lbMonitorResp models.CreateLBMonitorReq
 	if err := tftags.Get(d, &lbMonitorResp); err != nil {
 		return err
 	}
 
-	lbDetails, err := lb.lbClient.GetLoadBalancers(ctx)
```
Comment (gandharvas) [inline internal/cmp/lb_monitors.go:32]: Just for my understanding, why was this here before? Why has it been removed?

File: internal/cmp/lb_profiles.go [hunk 7]
```diff
@@ -115,15 +99,69 @@ func (lb *loadBalancerProfile) Create(ctx context.Context, d *utils.Data, meta i
...
+			ProfileConfig: models.LBProfile{
+				ProfileType:            d.GetString("profile_type"),
+				RequestHeaderSize:      d.GetInt("request_header_size"),
+				ResponseHeaderSize:     d.GetInt("response_header_size"),
+				ResponseTimeout:        d.GetInt("response_timeout"),
+				HTTPIdleTimeout:        d.GetInt("http_idle_timeout"),
+				FastTCPIdleTimeout:     d.GetInt("fast_tcp_idle_timeout"),
+				ConnectionCloseTimeout: d.GetInt("connection_close_timeout"),
+				HaFlowMirroring:        d.GetBool("ha_flow_mirroring"),
+				CookieMode:             d.GetString("cookie_mode"),
```
Comment (gandharvas) [inline internal/cmp/lb_profiles.go:138]: are these cookie related attributes collected from users as input?

File: internal/cmp/loadbalancer_datasource.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+	if err := d.Error(); err != nil {
+		return err
+	}
+
+	lb, err := n.lbClient.GetLoadBalancers(ctx)
+	if err != nil {
+		return err
+	}
+
+	for i, n := range lb.GetNetworkLoadBalancerResp {
```
Comment (gandharvas) [inline internal/cmp/loadbalancer_datasource.go:38]: I hope the loadblancer list response is not paginated. Otherwise, I hope the pagination is handled either in the library or in program logic

File: internal/resources/resource_load_balancer.go [hunk 9]
```diff
@@ -75,17 +75,19 @@ func LoadBalancer() *schema.Resource {
...
 							Type:        schema.TypeString,
 							Optional:    true,
+							Default:     "INFO",
 							Description: `In Filter. Supported Values are "DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL", "ALERT", "EMERGENCY"`,
 						},
 						"tier1": {
 							Type:        schema.TypeString,
 							Optional:    true,
-							Default:     true,
+							Default:     "/infra/tier-1s/26cdb82e-0057-4461-ad4d-cddd61d77b1f",
```
Comment (gandharvas) [inline internal/resources/resource_load_balancer.go:90]: Will this be static for all the environments?

Comment (Pushpalatha555) [reply internal/resources/resource_load_balancer.go:90]: yes

File: internal/cmp/lb_monitors.go (hunk 6 above)
Comment (Pushpalatha555) [reply internal/cmp/lb_monitors.go:32]: Before I have not used datasource to pick the loadBalancer details now i have used datasource from there i can able to create its configurations.

File: internal/cmp/constants.go (hunk 5 above)
Comment (Pushpalatha555) [reply internal/cmp/constants.go:32]: yes its case sensitive

File: examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf (hunk 4 above)
Comment (Pushpalatha555) [reply examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf:11]: yes its case sensitive i will update

File: internal/cmp/loadbalancer_datasource.go (hunk 8 above)
Comment (Pushpalatha555) [reply internal/cmp/loadbalancer_datasource.go:38]: its handled in GO SDK repo https://github.com/HewlettPackard/hpegl-vmaas-cmp-go-sdk/blob/254ea8100f702ee8640ccb920a9ddb77baa9e465/pkg/models/load_balancer.go#L101

File: internal/cmp/lb_profiles.go (hunk 7 above)
Comment (Pushpalatha555) [reply internal/cmp/lb_profiles.go:138]: cookie mode input is optional by default it will take INSERT if we want to change it as PREFIX, REWRITE then user needs to give the value

File: examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf (hunk 4 above)
Comment (gandharvas) [reply examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf]: Do we need validation or parsing for safety?

File: internal/cmp/constants.go (hunk 5 above)
Comment (gandharvas) [reply internal/cmp/constants.go:32]: Same here

File: internal/cmp/lb_profiles.go (hunk 7 above)
Comment (gandharvas) [reply internal/cmp/lb_profiles.go:138]: What about other attributes related to cookies

File: examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf (hunk 4 above)
Comment (Pushpalatha555) [reply examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf]: validation already added and we will get the error  "msg": "Field level validation errors: {value small of property size is not one of the allowed values [SMALL, MEDIUM, LARGE, XLARGE, DLB]}"

File: internal/cmp/lb_profiles.go (hunk 7 above)
Comment (Pushpalatha555) [reply internal/cmp/lb_profiles.go:138]: cookie_type, cookie_fallback, cookie_garbling and cookie_mode having default values but  cookie_name is required filed if we select cookie from serviceType

File: internal/resources/resource_load_balancer.go (hunk 9 above)
Comment (Pushpalatha555) [reply internal/resources/resource_load_balancer.go:90]: currently we have tier1 supporting for creating loadbalancer and its default one

File: internal/cmp/constants.go (hunk 5 above)
Comment (Pushpalatha555) [reply internal/cmp/constants.go:32]: validation already added and we will get the error  "msg": "Field level validation errors: {value Small of property size is not one of the allowed values [SMALL, MEDIUM, LARGE, XLARGE, DLB]}"

Comment (ravikumar29524) [inline internal/cmp/constants.go:32]: Please, if possible can we have a prefix of `default` for the `size` and `loglevel` , so that it is more readable.

**Review by ravikumar29524 — DISMISSED** [review]
LGTM. 
Please, as asked by Gandharva, get the tier1 default value confirmed from Manju

File: examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf [hunk 10]
```diff
@@ -1,15 +1,16 @@
...
+  enabled      =    true      
   resource_permission {
     all = true
   }
   config {
+    log_level = "INFO"
+    size = "SMALL"
     admin_state = true
-    tier1  = "/infra/tier-1s/26cdb82e-0057-4461-ad4d-cddd61d77b1f"
+    tier1 = "/infra/tier-1s/26cdb82e-0057-4461-ad4d-cddd61d77b1f"
```
Comment (manjunath-batakurki) [inline examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf:14]: Use a Data Source here

File: examples/resources/hpegl_vmaas_load_balancer_pool/nsx_t_lb_pool.tf [hunk 11]
```diff
@@ -1,6 +1,7 @@
 # (C) Copyright 2022 Hewlett Packard Enterprise Development LP
 
 resource "hpegl_vmaas_load_balancer_pool" "test_lb_pool" {
+  lb_id = data.hpegl_vmaas_load_balancer.lb.id
```
Comment (manjunath-batakurki) [inline examples/resources/hpegl_vmaas_load_balancer_pool/nsx_t_lb_pool.tf:4]: How can the user get these two values ?
 passive_monitor_path = 136
active_monitor_paths = 133

IN UI its a dropdown with some string values.
image

Comment (manjunath-batakurki) [reply examples/resources/hpegl_vmaas_load_balancer_pool/nsx_t_lb_pool.tf:4]: I also have a comment on the resource definition here. I assume it should be similar to UI so that its easy for end user in terms of End Usage.
image

@gandharvas  Need your input here

Comment (manjunath-batakurki) [reply examples/resources/hpegl_vmaas_load_balancer_pool/nsx_t_lb_pool.tf:4]: vip_balance = "ROUND_ROBIN" should have naming similar to UI

Comment (manjunath-batakurki) [reply examples/resources/hpegl_vmaas_load_balancer_pool/nsx_t_lb_pool.tf:4]: snat_ip_address = "" field in the UI. do we need this input from user ?

Comment (manjunath-batakurki) [reply examples/resources/hpegl_vmaas_load_balancer_pool/nsx_t_lb_pool.tf:4]: member_group  is dropdown in UI. why so many fields from user here ?

Comment (manjunath-batakurki) [reply examples/resources/hpegl_vmaas_load_balancer_pool/nsx_t_lb_pool.tf:4]: I see tags and scope in UI, don't we need them here ?

Comment (gandharvas) [reply examples/resources/hpegl_vmaas_load_balancer_pool/nsx_t_lb_pool.tf:4]: > I also have a comment on the resource definition here. I assume it should be similar to UI so that its easy for end user in terms of End Usage. image
> 
> @gandharvas Need your input here

I totally agree with you Manju, Any deviation from UI fields should be strictly restricted to mandatory fields only.(Which we cannot auto-populate)

File: examples/resources/hpegl_vmaas_load_balancer_profile/nsx_t_lb_profile.tf [hunk 12]
```diff
@@ -1,6 +1,7 @@
 # (C) Copyright 2022 Hewlett Packard Enterprise Development LP
 
 resource "hpegl_vmaas_load_balancer_profile" "tf_lb_profile" {
+  lb_id = data.hpegl_vmaas_load_balancer.lb.id
```
Comment (manjunath-batakurki) [inline examples/resources/hpegl_vmaas_load_balancer_profile/nsx_t_lb_profile.tf:4]: For the  "profile_type = "application-profile", do we need cookie details ? I think the example is not right. for the profileType= Persistance, we would need the right cookie details

Comment (manjunath-batakurki) [reply examples/resources/hpegl_vmaas_load_balancer_profile/nsx_t_lb_profile.tf:4]: Same comment on the resource definition. Completely different from UI
image

File: examples/resources/hpegl_vmaas_load_balancer_profile/nsx_t_lb_profile.tf [hunk 13]
```diff
@@ -1,6 +1,7 @@
 # (C) Copyright 2022 Hewlett Packard Enterprise Development LP
 
 resource "hpegl_vmaas_load_balancer_profile" "tf_lb_profile" {
+  lb_id = data.hpegl_vmaas_load_balancer.lb.id  
   name  =  "LB-PROFILE"       
   description  = "creating LB Profile"
   service_type     = "LBHttpProfile"
```
Comment (manjunath-batakurki) [inline examples/resources/hpegl_vmaas_load_balancer_profile/nsx_t_lb_profile.tf:7]: based on profile_type, the below fields changes for the user. It is highly diffficult for the end user to understand a clubbed resource definition to configure a given type of profile type. My suggestion would be separate it out for each profile. Currently for HTTP Service Type, there are more fields which are displayed in the UI, but not used in IaC.

The best way going forward is to analyse the resource definition before implementing it and see how can it be made helpful for the user consumption. 

Please refer Tier0/Tier1 Router creation for your reference.

I would suggest to take each of these Pools, Profiles and Monitors and Servers as independent stories so that we implement them in a right way.

Please let us know if we need a meeting to understand this better.

File: examples/resources/hpegl_vmaas_load_balancer_virtual_server/nsx_t_lb_virtual_server.tf [hunk 14]
```diff
@@ -2,12 +2,11 @@
 
 
 resource "hpegl_vmaas_load_balancer_virtual_server" "tf_lb_virtual_server" {
-  vip_name  =  "virtual"       
+  lb_id = data.hpegl_vmaas_load_balancer.lb.id
```
Comment (manjunath-batakurki) [inline examples/resources/hpegl_vmaas_load_balancer_virtual_server/nsx_t_lb_virtual_server.tf:5]: Same comment for the this one as well:
image

File: examples/resources/hpegl_vmaas_load_balancer_virtual_server/nsx_t_lb_virtual_server.tf [hunk 15]
```diff
@@ -2,12 +2,11 @@
 
 
 resource "hpegl_vmaas_load_balancer_virtual_server" "tf_lb_virtual_server" {
-  vip_name  =  "virtual"       
+  lb_id = data.hpegl_vmaas_load_balancer.lb.id 
+  vip_name  =  "virtual-server"       
   description  = "creating load balancer virtual server"
   vip_address     = "10.11.12.13"
-  vip_protocol = "http"
```
Comment (manjunath-batakurki) [inline examples/resources/hpegl_vmaas_load_balancer_virtual_server/nsx_t_lb_virtual_server.tf:8]: why is vip_protocol being removed ?

File: examples/resources/hpegl_vmaas_load_balancer_virtual_server/nsx_t_lb_virtual_server.tf [hunk 16]
```diff
@@ -2,12 +2,11 @@
...
   vip_port = "80"
-  pool = 306
```
Comment (manjunath-batakurki) [inline examples/resources/hpegl_vmaas_load_balancer_virtual_server/nsx_t_lb_virtual_server.tf:10]: pool should be a data source.

File: examples/resources/hpegl_vmaas_load_balancer_virtual_server/nsx_t_lb_virtual_server.tf [hunk 17]
```diff
@@ -2,12 +2,11 @@
...
   ssl_cert = 0 
   ssl_server_cert = 0
```
Comment (manjunath-batakurki) [inline examples/resources/hpegl_vmaas_load_balancer_virtual_server/nsx_t_lb_virtual_server.tf:11]: I exactly do not know important of these ssl_cert and ssl_server_cert. but i see SSL client profile being listed when we select these. Please discussion this with Jinu.

File: examples/resources/hpegl_vmaas_load_balancer_monitor/nsx_t_lb_monitor.tf [hunk 18]
```diff
@@ -1,7 +1,8 @@
 # (C) Copyright 2022 Hewlett Packard Enterprise Development LP
 
 resource "hpegl_vmaas_load_balancer_monitor" "tf_lb_monitor" {
-  name  =  "TEST-MONITOR"       
+  name  =  "TEST-MONITOR"  
+  lb_id = data.hpegl_vmaas_load_balancer.lb.id     
   description  = "Creating lb monitor"
   monitor_type = "LBTcpMonitorProfile"
```
Comment (manjunath-batakurki) [inline examples/resources/hpegl_vmaas_load_balancer_monitor/nsx_t_lb_monitor.tf:7]: based on the monitor type, the payload changes in UI. Need to see how can this implemented better.

**Review by manjunath-batakurki — CHANGES_REQUESTED** [review]
Hi Pushpalatha,

I think each of these resource definitions needs a change in the way they are implemented. I would suggest to setup a meeting. 
Points to discuss:
1. Resource definitions for Pools, Virtual Servers, Monitors and Profiles.
2. Story breakdown for individual resources
3. Any doubts you have regarding the implementation.

CC : @gandharvas
